### PR TITLE
Fill in some gaps in Strings documentation

### PR DESCRIPTION
**Pull request by benharsh**
* Makes the header more useful from a user perspective
* Indicate which functions may halt
* Add some code examples
* Adds documentation for some operator functions
* Wordsmithing

### CODE REVIEW

File: modules/internal/String.chpl [hunk 1]
```diff
@@ -155,9 +157,9 @@ module String {
     }
 
     /*
-      Construct a new string from the `c_string` `cs`. If `owned` is set to true, the
-      backing buffer will be freed when the new record is destroyed. If
-      `needToCopy` is set to true, the `c_string` will be copied into the
+      Construct a new string from the `c_string` `cs`. If `owned` is set to
```
Comment (lydia-duncan) [inline modules/internal/String.chpl]: Should we link to the c_string type here?

File: modules/internal/String.chpl [hunk 2]
```diff
@@ -362,11 +381,13 @@ module String {
...
     /*
-      Slice a string.
+      Slice a string. Halts if r is not completely inside the range
+      `1..string.length`.
 
       :arg r: range of the indices the new string should be made from
 
-      :returns: a substring within `1..string.length`
+      :returns: a new string that is a substring within `1..string.length`. If
+                `r.size` is zero, an empty string is returned.
```
Comment (lydia-duncan) [inline modules/internal/String.chpl]: Can we link to the size method on the range type?

File: modules/internal/String.chpl [hunk 3]
```diff
@@ -565,7 +586,8 @@ module String {
     /*
       :arg needle: the string to search for
       :arg region: an optional range defining the substring to search within,
-                   default is the whole string
+                   default is the whole string. Halts if range is not within
```
Comment (lydia-duncan) [inline modules/internal/String.chpl]: "if range" -> "if the range"

File: modules/internal/String.chpl [hunk 4]
```diff
@@ -590,7 +613,8 @@ module String {
...
+                   `1..string.length`
```
Comment (lydia-duncan) [inline modules/internal/String.chpl]: If you want to follow my last comment, here are two other spots to do it.

File: modules/internal/String.chpl [hunk 5]
```diff
@@ -1406,6 +1462,9 @@ module String {
   //
   // Append
   //
+  /*
+     Appends the string `lhs` with the string `rhs`.
```
Comment (lydia-duncan) [inline modules/internal/String.chpl]: I'd prefer this written as "Appends the string 'rhs' to the string 'lhs'." as the action is more direct then.

**Comment by lydia-duncan** [comment]
:+1:

File: modules/internal/String.chpl (hunk 1 above)
Comment (benharsh) [inline modules/internal/String.chpl]: I believe this would result in a broken link.

**Pull request merged**